File: stress/metrics/arr.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, List, Optional, Tuple

from OCRB.measure.events import Event, EventType, FailureClass


@dataclass(frozen=True)
class ARRResult:
    arr: Optional[float]          # None = N/A
    Fr: int                       # recoverable failures observed
    Fa: int                       # autonomously recovered failures observed
    na_reason: Optional[str] = None


def compute_arr(events: List[Event]) -> ARRResult:
    """
    BP-2 — Autonomous Recovery Rate (ARR)
    ARR = Fa / Fr

    Where:
      Fr = number of recoverable failures observed (incl. auto-recovered + not recovered)
      Fa = number of recoverable failures autonomously recovered

    Edge case:
      If Fr == 0, ARR is N/A (metric not applicable). Must be disclosed.
    """
    Fr = 0
    Fa = 0

    for e in events:
        if e.type == EventType.FAILURE and e.failure_class is not None:
            if e.failure_class in (
                FailureClass.AUTONOMOUSLY_RECOVERED,
                FailureClass.RECOVERABLE_NOT_RECOVERED,
            ):
                Fr += 1
                if e.failure_class == FailureClass.AUTONOMOUSLY_RECOVERED:
                    Fa += 1

    if Fr == 0:
        return ARRResult(arr=None, Fr=0, Fa=0, na_reason="Fr=0 (no recoverable failures observed)")

    return ARRResult(arr=Fa / Fr, Fr=Fr, Fa=Fa, na_reason=None)

File: stress-ref/src/metrics/arr.rs
use crate::types::events::{Event, EventType, FailureClass};

#[derive(Debug, Clone)]
pub struct ArrResult {
    pub arr: Option<f64>,
    pub fr: i64,
    pub fa: i64,
    pub na_reason: Option<String>,
}

/// BP-2: Autonomous Recovery Rate. ARR = Fa / Fr.
pub fn compute_arr(events: &[Event]) -> ArrResult {
    let mut fr: i64 = 0;
    let mut fa: i64 = 0;

    for e in events {
        if e.event_type == EventType::Failure {
            if let Some(fc) = &e.failure_class {
                match fc {
                    FailureClass::AutonomouslyRecovered => {
                        fr += 1;
                        fa += 1;
                    }
                    FailureClass::RecoverableNotRecovered => {
                        fr += 1;
                    }
                    FailureClass::Irreversible => {}
                }
            }
        }
    }

    if fr == 0 {
        return ArrResult {
            arr: None, fr: 0, fa: 0,
            na_reason: Some("Fr=0 (no recoverable failures observed)".to_string()),
        };
    }

    ArrResult {
        arr: Some(fa as f64 / fr as f64),
        fr, fa, na_reason: None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::events::EventLog;

    #[test]
    fn all_recovered() {
        let mut log = EventLog::new("t", "W1-A");
        log.emit(EventType::RunStart);
        let e = log.emit(EventType::Failure);
        e.failure_id = Some("f1".into());
        e.failure_class = Some(FailureClass::AutonomouslyRecovered);
        let e = log.emit(EventType::Failure);
        e.failure_id = Some("f2".into());
        e.failure_class = Some(FailureClass::AutonomouslyRecovered);
        log.emit(EventType::RunEnd);

        let result = compute_arr(log.events());
        assert_eq!(result.arr, Some(1.0));
    }

    #[test]
    fn partial_recovery() {
        let mut log = EventLog::new("t", "W1-A");
        log.emit(EventType::RunStart);
        let e = log.emit(EventType::Failure);
        e.failure_class = Some(FailureClass::AutonomouslyRecovered);
        let e = log.emit(EventType::Failure);
        e.failure_class = Some(FailureClass::RecoverableNotRecovered);
        log.emit(EventType::RunEnd);

        let result = compute_arr(log.events());
        assert_eq!(result.arr, Some(0.5));
    }

    #[test]
    fn no_failures_na() {
        let mut log = EventLog::new("t", "W1-A");
        log.emit(EventType::RunStart);
        log.emit(EventType::RunEnd);
        let result = compute_arr(log.events());
        assert!(result.arr.is_none());
    }
}
